core/src/main/scala/org/bitcoins/core/wallet/utxo/UTXOSpendingInfo.scala

```diff
@@ -17,26 +15,71 @@ sealed abstract class UTXOSpendingInfo {
   /** The funding transaction's txid and the index of the output in the transaction we are spending */
   def outPoint: TransactionOutPoint
 
+  def amount: CurrencyUnit
+
+  def scriptPubKey: ScriptPubKey
+
   /** the actual output itself we are spending */
-  def output: TransactionOutput
+  def output: TransactionOutput = {
+    TransactionOutput(value = amount,
+      scriptPubKey = scriptPubKey)
+  }
 
   /** the signers needed to spend from the output above */
   def signers: Seq[Sign]
 
-  /** a redeemScript, if required, to spend the output above */
-  def redeemScriptOpt: Option[ScriptPubKey]
+  def hashType: HashType
+}
 
-  /** the scriptWitness, if required, to spend the output above */
-  def scriptWitnessOpt: Option[ScriptWitness]
+sealed trait BitcoinUTXOSpendingInfo extends UTXOSpendingInfo
 
-  def hashType: HashType
+/** This represents the information needed to be spend scripts like
+  * [[org.bitcoins.core.protocol.script.P2PKHScriptPubKey p2pkh]] or [[org.bitcoins.core.protocol.script.P2PKScriptPubKey p2pk]]
+  * scripts. Basically there is no nesting that requires a redeem script here*/
+case class RawScriptUTXOSpendingInfo(outPoint: TransactionOutPoint,
+                                     amount: CurrencyUnit,
+                                     scriptPubKey: ScriptPubKey,
+                                     signers: Seq[Sign],
+                                     hashType: HashType) extends BitcoinUTXOSpendingInfo
+
+
+/** This is the case where we are spending a [[org.bitcoins.core.protocol.script.WitnessScriptPubKeyV0 witness v0 script]]  */
+case class SegwitV0NativeUTXOSpendingInfo(outPoint: TransactionOutPoint,
+                                          amount: CurrencyUnit,
+                                          scriptPubKey: WitnessScriptPubKeyV0,
```

Comment:
I think the most important things here are the tightenting of types on scripts, for instance `WitnessScriptPubKeyV0` here rather than just `ScriptPubKey`. This should reduce the amount of pattern matching required significantly.